Avoid dbAsyncDelete() in performEvictions() even if lazy_eviction is set

moticless opened the pull request:
Avoid dbAsyncDelete() in performEvictions(). Even if lazy_eviction is configured, it
is inefficient to lazy-free and let the main thread just to wait to memory reduction

oranagra commented:
@moticless please add some details on what you measured.
have you considered an immense object being evicted (e.g. a hash with 100k fields)?

moticless commented:
In general, when lazy-free is configured,  the main thread calls dbAsyncDelete() and waits … it is lazy handled by the lazy-free background job which in turn requests jemalloc-thread to free it. All this flow is very inefficient.

To verify my findings, as performEvictions() has single entry and exit point, I added the following two printouts to measure amount of time spent in performEvictions once it is identified that we above maxmemory, until the point reached below it:


   
At start:
```
   static int _state=0;
   static long long measured_time;
    if (_state == 0) { 
      measured_time = ustime(); 
      printf("performEviction() start time (usec): %llu\n",  measured_time); 
      _state = 1; 
   }
```

At then end:
```
    if (result == EVICT_OK) {
        if (_state == 1) {
            printf("Measured duration: %llu\n", ustime() -  measured_time);
            _state = 0; 
        }
    } else {
        printf("performEviction() returned result other than EVICT_OK (=%d)\n", result);
    }
```
    

Then i configured policy to allkeys-lru and added 10 lists, 100000 items each. In total ~48MB. Then I run the command config set maxmemory 28216568. And got the following interesting printouts:

lazyfree-lazy-eviction yes:  Measured duration: 329984

lazyfree-lazy-eviction no: Measured duration: 9464

Note: In both cases they completely removed all the 10 lists, even though maxmemory is set 28216568. This is because jemalloc thread will release the allocations on a distinct thread.

moticless commented:
@oranagra , On performEvictions(), I wonder if there is someway to lock jemalloc and delete directly the objects by the main thread. In such extreme scenario of crossing maxmemory, we might gain accurate release without redundant evictions, and faster roundtrip.

soloestoy commented:
>lazyfree-lazy-eviction yes: Measured duration: 329984
>lazyfree-lazy-eviction no: Measured duration: 9464

you can try it again after #11237 merged.

and I don't think disable lazyfree-eiviction is right, it's useful to avoid latency spike cased by big key evicted. But we can reduce the influence by lazyfree in eviction, like wait lazyfree before do eviction, see more details in #11294

moticless commented:
Hi @soloestoy, I probably was too hasty about avoiding lazy-eviction (which I still think that in the common case, do bad job). I need to review again my measurements and also recent changes and see if i can refine my case of when to avoid lazy-eviction. Thank you.

close this PR.
